source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -0,0 +1,173 @@
+  if (config_->doRequest()) {
+    writeHeaderToMetadata(headers, config_->requestRules(), *decoder_callbacks_);
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void HeaderToMetadataFilter::setDecoderFilterCallbacks(
+    Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+Http::FilterHeadersStatus HeaderToMetadataFilter::encodeHeaders(Http::HeaderMap& headers, bool) {
+  if (config_->doResponse()) {
+    writeHeaderToMetadata(headers, config_->responseRules(), *encoder_callbacks_);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void HeaderToMetadataFilter::setEncoderFilterCallbacks(
+    Http::StreamEncoderFilterCallbacks& callbacks) {
+  encoder_callbacks_ = &callbacks;
+}
+
+bool HeaderToMetadataFilter::addMetadata(StructMap& map, const std::string& metaNamespace,
+                                         const std::string& key, const std::string& value,
+                                         ValueType type) const {
+  ProtobufWkt::Value val;
+
+  if (value.empty()) {
+    // No value, skip. we could allow this though.
+    ENVOY_LOG(debug, "no metadata value provided");
+    return false;
+  }
+
+  if (value.size() >= MAX_HEADER_VALUE_LEN) {
+    // Too long, go away.
+    ENVOY_LOG(debug, "metadata value is too long");
+    return false;
+  }
+
+  // Sane enough, add the key/value.
+  switch (type) {
+  case envoy::config::filter::http::header_to_metadata::v2::Config_ValueType_STRING:
+    val.set_string_value(value);
+    break;
+  case envoy::config::filter::http::header_to_metadata::v2::Config_ValueType_NUMBER:
+    try {
+      val.set_number_value(std::stod(value));
+    } catch (...) {
+      ENVOY_LOG(debug, "value to number conversion failed");
+      return false;
+    }
+    break;
+  default:
+    ENVOY_LOG(debug, "unknown value type");
+    return false;
+  }
+
+  // Have we seen this namespace before?
+  auto namespaceIter = map.find(metaNamespace);
+  if (namespaceIter == map.end()) {
+    map[metaNamespace] = ProtobufWkt::Struct();
+    namespaceIter = map.find(metaNamespace);
+  }
+
+  auto& keyval = namespaceIter->second;
+  (*keyval.mutable_fields())[key] = val;
+
+  return true;
+}
+
+const std::string& HeaderToMetadataFilter::decideNamespace(const std::string& nspace) const {
+  return nspace.empty() ? HttpFilterNames::get().HEADER_TO_METADATA : nspace;
+}
+
+void HeaderToMetadataFilter::writeHeaderToMetadata(Http::HeaderMap& headers,
+                                                   const HeaderToMetadataRules& rules,
+                                                   Http::StreamFilterCallbacks& callbacks) {
+  StructMap structsByNamespace;
+
+  for (const auto& rulePair : rules) {
+    const auto& header = rulePair.first;
+    const auto& rule = rulePair.second;
+    const Http::HeaderEntry* header_entry = headers.get(header);
+
+    if (header_entry != nullptr) {
+      const auto& keyval = rule.on_header_present();
```

Comment:
I think something needs to validate that on_header_present was actually set. I believe this can be done with annotations on the protobuf, but I'm not sure of the exact syntax.